service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/MedicationStatementExtractor.java

```diff
@@ -0,0 +1,129 @@
+package uk.nhs.adaptors.gp2gp.ehr.mapper;
+
+import org.apache.commons.lang3.StringUtils;
+import org.hl7.fhir.dstu3.model.CodeableConcept;
+import org.hl7.fhir.dstu3.model.DateTimeType;
+import org.hl7.fhir.dstu3.model.MedicationRequest;
+import org.hl7.fhir.dstu3.model.Reference;
+import org.hl7.fhir.dstu3.model.codesystems.MedicationRequestIntent;
+
+import com.github.mustachejava.Mustache;
+
+import uk.nhs.adaptors.gp2gp.ehr.exception.EhrMapperException;
+import uk.nhs.adaptors.gp2gp.ehr.mapper.parameters.InFulfilmentOfTemplateParameters;
+import uk.nhs.adaptors.gp2gp.ehr.utils.DateFormatUtil;
+import uk.nhs.adaptors.gp2gp.ehr.utils.TemplateUtils;
+
+public class MedicationStatementExtractor {
+    private static final Mustache IN_FULFILMENT_OF_TEMPLATE =
+        TemplateUtils.loadTemplate("in_fulfilment_of_template.mustache");
+
+    private static final String MEDICATION_QUANTITY_TEXT = "https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-MedicationQuantityText-1";
+    private static final String PRESCRIPTION_TYPE_URL = "https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-PrescriptionType-1";
+    private static final String REPEAT_INFORMATION_URL = "https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-MedicationRepeatInformation-1";
+    private static final String NUM_OF_REPEAT_PRESCRIPTIONS_ALLOWED_URL = "numberOfRepeatPrescriptionsAllowed";
+    private static final String DEFAULT_REPEAT_VALUE = "1";
+    private static final String MEDICATION_STATUS_REASON_URL = "https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-MedicationStatusReason-1";
+    private static final String STATUS_REASON_URL = "statusReason";
+    private static final String STATUS_CHANGE_URL = "statusChangeDate";
+    private static final String AVAILABILITY_TIME_VALUE_TEMPLATE = "<availabilityTime value=\"%s\"/>";
+    private static final String DEFAULT_QUANTITY_TEXT = "Unk UoM";
+
+    public static String extractDispenseRequestQuantityText(MedicationRequest medicationRequest) {
+        return medicationRequest.getDispenseRequest()
+            .getExtension()
+            .stream()
+            .filter(value -> value.getUrl().equals(MEDICATION_QUANTITY_TEXT))
+            .findFirst()
+            .map(value -> value.getValue().toString())
+            .orElse(DEFAULT_QUANTITY_TEXT);
+    }
+
+    public static String extractPrescriptionTypeCode(MedicationRequest medicationRequest) {
+        return medicationRequest.getExtension()
+            .stream()
+            .filter(value -> value.getUrl().equals(PRESCRIPTION_TYPE_URL))
+            .findFirst()
+            .map(value -> (CodeableConcept) value.getValue())
+            .map(code -> code.getCodingFirstRep().getCode())
+            .orElse(StringUtils.EMPTY);
+    }
+
+    public static String extractRepeatValue(MedicationRequest medicationRequest) {
+        var repeatInformation = medicationRequest.getExtension()
```

Comment:
There is a utility called ExtensionMappingUtil that will take a resource and find an extension based on URL, might reduce a little of the code.